File: Data_structure/837/num_of_points.cpp
#include <iostream>
using namespace std;

const int N = 1e5 + 10;
int n, m, a, b, pa, pb, p[N], cnt[N];
char op[3];

int find(int x);

int main() {
    scanf("%d%d", &n, &m);
    for (int i = 1; i <= n; i++) {
        p[i] = i;
        cnt[i] = 1;
    }
    while (m--) {
        scanf("%s", op);
        if (op[0] == 'C') {
            scanf("%d%d", &a, &b);
            pa = find(a), pb = find(b);
            if (pa != pb) {
                p[pa] = pb;
                cnt[pb] += cnt[pa];
            }
        } else if (op[1] == '1') {
            scanf("%d%d", &a, &b);
            pa = find(a), pb = find(b);
            if (pa == pb) printf("Yes\n");
            else printf("No\n");
        } else {
            scanf("%d", &a);
            pa = find(a);
            printf("%d\n", cnt[pa]);
        }
    }
    return 0;
}

int find(int x) {
    if (p[x] != x) p[x] = find(p[x]);
    return p[x];
}

File: Data_structure/837/num_of_points.py
def find(x: int, p: list) -> int:
    root = x
    while root != p[root]:
        root = p[root]
    while x != p[x]:
        t = p[x]
        p[x] = root
        x = t
    return root

if __name__ == "__main__":
    n, m = map(int, input().split())
    p = [i for i in range(n + 1)]
    cnt = [1] * (n + 1)
    for _ in range(m):
        line = input().split()
        op = line[0]
        if op == "C":
            x, y = int(line[1]), int(line[2])
            px, py = find(x, p), find(y, p)
            if px != py:
                p[px] = py
                cnt[py] += cnt[px]
        elif op == "Q1":
            x, y = int(line[1]), int(line[2])
            px, py = find(x, p), find(y, p)
            if px == py:
                print("Yes")
            else:
                print("No")
        else:
            x = int(line[1])
            print(cnt[find(x, p)])
